### PR TITLE
Integration of GRIDE method

### PR DESCRIPTION
**Pull request by Malinon**

### CODE REVIEW

File: skdim/id/_GRIDE.py [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+        if type(n1) != int or type(self.n2) != int:
+            raise ValueError("n1 and n2 must be integers")
+        if n1 < 1 or self.n2 < 1:
+            raise ValueError("n1 and n2 must be greater than 0")
+        if d0 > d1:
+            raise ValueError("Interval is not proper. d0 must be smaller than d1")
+        if d0 < 0 or d1 < 0:
+            raise ValueError("Dimenstions d0 and d1 must be greater than 0")
+    
+    def fit(self, X, y=None, n_jobs=1):
```
Comment (kmyim) [inline skdim/id/_GRIDE.py:107]: move `n_jobs` to argument of `__init__`?

File: skdim/id/_GRIDE.py [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+        ----------
+        X : {array-like}, shape (n_samples, n_features)
+            A data set for which the intrinsic dimension is estimated.
+        y : dummy parameter to respect the sklearn API
+
+        Returns
+        self : object
+            Returns self.
+        """
+        self.n_jobs = n_jobs
```
Comment (kmyim) [inline skdim/id/_GRIDE.py:121]: Set `n_jobs` in `__init__`?

File: skdim/id/_GRIDE.py [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+
+
+class Gride(GlobalEstimator):
+    """Intrinsic dimension estimation using the GRIDE algorithm
+        References:
+        F. Denti, D. Doimo, A. Laio, A. Mira, Distributional results for model-based intrinsic dimension
+            estimators, arXiv preprint arXiv:2104.13832 (2021).
+    """
+    
+    def __init__(self, n1=1, n2=None, d0=0.001, d1=1000, eps=1e-7, range_max=None, metric="euclidean"):
```
Comment (kmyim) [inline skdim/id/_GRIDE.py:66]: Move `n_jobs` from `fit` to `__init__`?

File: skdim/tests/global/gride_test.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    np.random.seed(782)
+    X = np.random.normal(0, 1, (1000, 3))
+    data = Data(X)
+    grid = Gride(range_max=64)
+    grid.fit(X)
+    dim_evolution = grid.transform_multiscale()
+    id_list, _, _ = data.return_id_scaling_gride(range_max=64)
+    assert np.allclose(id_list, dim_evolution, atol=0.1)
+
+def test_avg_results_for_hypercube():
```
Comment (kmyim) [inline skdim/tests/global/gride_test.py:67]: This test is taking too long!! Computation too large, consider reducing `SAMPLE_SIZE` or `MONTE_CARLO_SIMULATION_NUM`?

File: skdim/tests/global/gride_test.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    # Values are taken from original paper, Figure 4
+    MONTE_CARLO_SIMULATION_NUM = 50
+    SAMPLE_SIZE = 10000
+    N1 = 100
+    real_dims = [2 * i for i in range(1, 6)]
+    hypercubes = [np.random.uniform(low=0.0, high=1.0, size=(SAMPLE_SIZE, dim)) for dim in real_dims]
```
Comment (kmyim) [inline skdim/tests/global/gride_test.py:73]: Redundant, defined in loop below

File: skdim/id/_gride_external_function.py
Comment (kmyim) [inline skdim/id/_gride_external_function.py:1]: Do we need this?

**Review by kmyim — COMMENTED** [review]
Looks good, but the Monte Carlo simulation test takes too long to run on my laptop and couldn't finish the test in reasonable time?

File: skdim/tests/global/gride_test.py (hunk 4 above)
Comment (Malinon) [reply skdim/tests/global/gride_test.py:67]: It is test that is run once or twice per GRIDE method integration. Excpected results are taken from original paper so it is required to use the same parameters as in original paper.

This test will be run only few times. As long as the test if finished in finite time, it is ok in my opinion :)

File: skdim/id/_GRIDE.py (hunk 3 above)
Comment (Malinon) [reply skdim/id/_GRIDE.py]: Done in newest commit

File: skdim/id/_GRIDE.py (hunk 1 above)
Comment (Malinon) [reply skdim/id/_GRIDE.py]: Done in newest commit

File: skdim/id/_GRIDE.py [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+        """Compute the id at different scales using the Gride algorithm.
+
+        Args:
+            X (np.ndarray(float)): data matrix of size (n_samples, n_features)
+        Returns:
+            ids_scaling (np.ndarray(float)): array of intrinsic dimensions of length log2(range_max);
+        """
+
+        max_step = int(math.log(self.range_max, 2))
+        nn_ranks = np.array([2**i for i in range(max_step + 1)])
```
[the remark this answers is not in the document]
Comment (Malinon) [reply skdim/id/_GRIDE.py]: Removed in newest commit

File: skdim/tests/global/gride_test.py (hunk 5 above)
Comment (Malinon) [reply skdim/tests/global/gride_test.py:73]: Removed in newest commit

**Comment by Malinon** [comment]
I update code based on review

**Review by kmyim — APPROVED** [review]

**Pull request merged**